pkg/helmshim/helm.go

```diff
@@ -50,22 +51,75 @@ func isNullOrEmptyObject(o *unstructured.Unstructured) bool {
 	return string(b) == "null" || string(b) == "{}"
 }
 
-func RenderChart(opts ...RenderOption) error {
-	a := action.NewInstall(&action.Configuration{})
-	a.ReleaseName = "eno-helm-shim"
-	a.Namespace = "default"
-	a.DryRun = true
-	a.Replace = true
-	a.ClientOnly = true
-	a.IncludeCRDs = true
-
-	o := &options{
-		Action:      a,
-		ValuesFunc:  inputsToValues,
-		ChartLoader: defaultChartLoader,
+// Synth produced a SynthFunc that 1) uses inputs as a values func ratehr than input reader 2) a writer that saves the objects to a slice rather than serialing
+func Synth[T function.Inputs](values func(inputs T) (map[string]any, error), opts ...RenderOption) function.SynthFunc[T] {
+
+	return func(inputs T) ([]client.Object, error) {
+
+		a := action.NewInstall(&action.Configuration{})
+		a.ReleaseName = "eno-helm-shim"
+		a.Namespace = "default"
+		a.DryRun = true
+		a.Replace = true
+		a.ClientOnly = true
+		a.IncludeCRDs = true
+
+		o := &options{
+			Action:      a,
+			ChartLoader: defaultChartLoader,
+		}
+		for _, opt := range opts {
+			opt.apply(o)
+		}
+
+		c, err := o.ChartLoader()
+		if err != nil {
+			return nil, errors.Join(ErrChartNotFound, err)
+		}
+
+		//error if reader is set?
+
+		vals, err := values(inputs)
+		if err != nil {
+			return nil, errors.Join(ErrConstructingValues, err)
+		}
+
+		rel, err := a.Run(c, vals)
+		if err != nil {
+			return nil, errors.Join(ErrRenderAction, err)
+		}
+
+		b := bytes.NewBufferString(rel.Manifest)
+		// append manifest from hook
+		for _, hook := range rel.Hooks {
+			fmt.Fprintf(b, "---\n# Source: %s\n%s\n", hook.Name, hook.Manifest)
+		}
+
+		var results []client.Object
+		d := yaml.NewYAMLToJSONDecoder(b)
+		for {
+			m := &unstructured.Unstructured{}
+			err = d.Decode(m)
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, errors.Join(ErrCannotParseChart, err)
+			}
+			if isNullOrEmptyObject(m) {
+				continue
+			}
+			results = append(results, m.DeepCopy())
```

Comment:
Do we need to copy here?

Reply:
was mimicing @tyler-lloyd  I don't think so but do have to cast?

Reply:
tests say no.